File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/model/orderstatus/TradeItOrderLegParcelable.java
package it.trade.android.sdk.model.orderstatus;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

import it.trade.model.reponse.Fill;
import it.trade.model.reponse.OrderLeg;

public class TradeItOrderLegParcelable implements Parcelable {

    @SerializedName("priceInfo")
    private TradeItPriceInfoParcelable priceInfo;

    @SerializedName("fills")
    private List<TradeItFillParcelable> fills = new ArrayList<>();

    @SerializedName("symbol")
    private String symbol;

    @SerializedName("orderedQuantity")
    private Integer orderedQuantity;

    @SerializedName("filledQuantity")
    private Integer filledQuantity;

    @SerializedName("action")
    private String action;

    TradeItOrderLegParcelable(OrderLeg orderLeg) {
        this.priceInfo = new TradeItPriceInfoParcelable(orderLeg.priceInfo);
        this.fills = mapFillToTradeItFillParcelable(orderLeg.fills);
        this.symbol = orderLeg.symbol;
        this.orderedQuantity = orderLeg.orderedQuantity;
        this.filledQuantity = orderLeg.filledQuantity;
        this.action = orderLeg.action;
    }

    public TradeItPriceInfoParcelable getPriceInfo() {
        return priceInfo;
    }

    public List<TradeItFillParcelable> getFills() {
        return fills;
    }

    public String getSymbol() {
        return symbol;
    }

    public Integer getOrderedQuantity() {
        return orderedQuantity;
    }

    public Integer getFilledQuantity() {
        return filledQuantity;
    }

    public String getAction() {
        return action;
    }

    @Override
    public String toString() {
        return "TradeItOrderLegParcelable{" +
                "priceInfo=" + priceInfo +
                ", fills=" + fills +
                ", symbol='" + symbol + '\'' +
                ", orderedQuantity=" + orderedQuantity +
                ", filledQuantity=" + filledQuantity +
                ", action='" + action + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        TradeItOrderLegParcelable that = (TradeItOrderLegParcelable) o;

        if (priceInfo != null ? !priceInfo.equals(that.priceInfo) : that.priceInfo != null)
            return false;
        if (!fills.equals(that.fills)) return false;
        if (symbol != null ? !symbol.equals(that.symbol) : that.symbol != null) return false;
        if (orderedQuantity != null ? !orderedQuantity.equals(that.orderedQuantity) : that.orderedQuantity != null)
            return false;
        if (filledQuantity != null ? !filledQuantity.equals(that.filledQuantity) : that.filledQuantity != null)
            return false;
        return action != null ? action.equals(that.action) : that.action == null;

    }

    @Override
    public int hashCode() {
        int result = priceInfo != null ? priceInfo.hashCode() : 0;
        result = 31 * result + fills.hashCode();
        result = 31 * result + (symbol != null ? symbol.hashCode() : 0);
        result = 31 * result + (orderedQuantity != null ? orderedQuantity.hashCode() : 0);
        result = 31 * result + (filledQuantity != null ? filledQuantity.hashCode() : 0);
        result = 31 * result + (action != null ? action.hashCode() : 0);
        return result;
    }

    private List<TradeItFillParcelable> mapFillToTradeItFillParcelable(List<Fill> fills) {
        List<TradeItFillParcelable> fillParcelableList = new ArrayList<>();
        for (Fill fill: fills) {
            fillParcelableList.add(new TradeItFillParcelable(fill));
        }
        return fillParcelableList;
    }


    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(this.priceInfo, flags);
        dest.writeTypedList(this.fills);
        dest.writeString(this.symbol);
        dest.writeValue(this.orderedQuantity);
        dest.writeValue(this.filledQuantity);
        dest.writeString(this.action);
    }

    protected TradeItOrderLegParcelable(Parcel in) {
        this.priceInfo = in.readParcelable(TradeItPriceInfoParcelable.class.getClassLoader());
        this.fills = in.createTypedArrayList(TradeItFillParcelable.CREATOR);
        this.symbol = in.readString();
        this.orderedQuantity = (Integer) in.readValue(Integer.class.getClassLoader());
        this.filledQuantity = (Integer) in.readValue(Integer.class.getClassLoader());
        this.action = in.readString();
    }

    public static final Parcelable.Creator<TradeItOrderLegParcelable> CREATOR = new Parcelable.Creator<TradeItOrderLegParcelable>() {
        @Override
        public TradeItOrderLegParcelable createFromParcel(Parcel source) {
            return new TradeItOrderLegParcelable(source);
        }

        @Override
        public TradeItOrderLegParcelable[] newArray(int size) {
            return new TradeItOrderLegParcelable[size];
        }
    };
}

File: tradeit-android-sdk/src/main/java/it/trade/android/sdk/model/orderstatus/TradeItOrderLegParcelable.kt
package it.trade.android.sdk.model.orderstatus

import android.os.Parcel
import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import it.trade.model.reponse.Fill
import it.trade.model.reponse.OrderLeg
import java.util.*

class TradeItOrderLegParcelable : Parcelable {

    @SerializedName("priceInfo")
    var priceInfo: TradeItPriceInfoParcelable? = null
        private set

    @SerializedName("fills")
    var fills: List<TradeItFillParcelable> = ArrayList()
        private set

    @SerializedName("symbol")
    var symbol: String? = null
        private set

    @SerializedName("orderedQuantity")
    var orderedQuantity: Int? = null
        private set

    @SerializedName("filledQuantity")
    var filledQuantity: Int? = null
        private set

    @SerializedName("action")
    var action: String? = null
        private set

    internal constructor(orderLeg: OrderLeg) {
        this.priceInfo = TradeItPriceInfoParcelable(orderLeg.priceInfo)
        this.fills = mapFillToTradeItFillParcelable(orderLeg.fills)
        this.symbol = orderLeg.symbol
        this.orderedQuantity = orderLeg.orderedQuantity
        this.filledQuantity = orderLeg.filledQuantity
        this.action = orderLeg.action
    }

    override fun toString(): String {
        return "TradeItOrderLegParcelable{" +
                "priceInfo=" + priceInfo +
                ", fills=" + fills +
                ", symbol='" + symbol + '\''.toString() +
                ", orderedQuantity=" + orderedQuantity +
                ", filledQuantity=" + filledQuantity +
                ", action='" + action + '\''.toString() +
                '}'.toString()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false

        val that = o as TradeItOrderLegParcelable?

        if (if (priceInfo != null) priceInfo != that!!.priceInfo else that!!.priceInfo != null)
            return false
        if (fills != that.fills) return false
        if (if (symbol != null) symbol != that.symbol else that.symbol != null) return false
        if (if (orderedQuantity != null) orderedQuantity != that.orderedQuantity else that.orderedQuantity != null)
            return false
        if (if (filledQuantity != null) filledQuantity != that.filledQuantity else that.filledQuantity != null)
            return false
        return if (action != null) action == that.action else that.action == null

    }

    override fun hashCode(): Int {
        var result = if (priceInfo != null) priceInfo!!.hashCode() else 0
        result = 31 * result + fills.hashCode()
        result = 31 * result + if (symbol != null) symbol!!.hashCode() else 0
        result = 31 * result + if (orderedQuantity != null) orderedQuantity!!.hashCode() else 0
        result = 31 * result + if (filledQuantity != null) filledQuantity!!.hashCode() else 0
        result = 31 * result + if (action != null) action!!.hashCode() else 0
        return result
    }

    private fun mapFillToTradeItFillParcelable(fills: List<Fill>): List<TradeItFillParcelable> {
        val fillParcelableList = ArrayList<TradeItFillParcelable>()
        for (fill in fills) {
            fillParcelableList.add(TradeItFillParcelable(fill))
        }
        return fillParcelableList
    }


    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(this.priceInfo, flags)
        dest.writeTypedList(this.fills)
        dest.writeString(this.symbol)
        dest.writeValue(this.orderedQuantity)
        dest.writeValue(this.filledQuantity)
        dest.writeString(this.action)
    }

    protected constructor(`in`: Parcel) {
        this.priceInfo = `in`.readParcelable(TradeItPriceInfoParcelable::class.java!!.getClassLoader())
        this.fills = `in`.createTypedArrayList(TradeItFillParcelable.CREATOR)
        this.symbol = `in`.readString()
        this.orderedQuantity = `in`.readValue(Int::class.java!!.getClassLoader()) as Int
        this.filledQuantity = `in`.readValue(Int::class.java!!.getClassLoader()) as Int
        this.action = `in`.readString()
    }

    companion object {

        val CREATOR: Parcelable.Creator<TradeItOrderLegParcelable> = object : Parcelable.Creator<TradeItOrderLegParcelable> {
            override fun createFromParcel(source: Parcel): TradeItOrderLegParcelable {
                return TradeItOrderLegParcelable(source)
            }

            override fun newArray(size: Int): Array<TradeItOrderLegParcelable?> {
                return arrayOfNulls(size)
            }
        }
    }
}
